### ISSUE TITLE
[Question] Set max steps per epoch? Calculation of `val_loss` during epochs?

### ISSUE THREAD

**Issue by gatoniel** [CAREamics/careamics#510]
## Question

Is there a way to set the steps per epoch? Or, is there a way to set after how many steps the validation loss should be calculated?

I ask for the following reason: (A) It seems that one epoch is defined as looping through all the possible training data once. (B) It seems that the validation loss `val_loss` is calculated only once per epoch at the very end of the epoch. (C) With `checkpoint_params={"every_n_train_steps": 100,},` checkpoints can be generated in between epochs, here after 100 steps. However it is unclear on which basis this checkpoint evaluates/chooses the models as there is no updated validation loss available in between epochs.

In my case, I have a big dataset of 200 GB that I would like to denoise with CAREamics N2V. Ideally, I would like to use the whole dataset for training (90 %) and validation (10 %). Given the big size of the dataset, one epoch has many steps and takes very long (< 1 epoch per hour). For a smaller dataset, I have ~70 epochs in two hours. For this smaller dataset, the final model was chosen from 70 different checkpoints (each at the end of one of the 70 epochs). However, for the big dataset, the final model was chosen only from 6 different checkpoints allthough a lot more training steps were performed. I would like to increase the "sampling rate". So, either I need to reduce the number of training steps per epoch (randomly some of the training data appears in an epoch, some does not) or I need to somehow evaluate the validation loss in between epochs.

Maybe, you have another suggestion?

EDIT:
I just found out that there is the [`val_check_interval`](https://lightning.ai/docs/pytorch/stable/common/trainer.html#val-check-interval) parameter in PytrochLightning. However, currently it cannot be set in the CAREamics configuration and it is [not passed to the `Trainer`](https://github.com/CAREamics/careamics/blob/9f14e63a5fa25dcc184cd4b67a6f9b181af7607d/src/careamics/careamist.py#L210). I think this would be the easiest solution for the above problem.

EDIT2:
The above "solution" is not optimal, as it would not yet change the behaviour of the learning rate scheduler. The lr_scheduler would still only be called at the end of epochs. Hence, the learning rate would not be tuned as fast as it would be tuned in a case with fewer steps per epoch. I think this is linked to issue https://github.com/CAREamics/careamics/issues/480.

**Comment by jdeschamps** [comment]
Hi @gatoniel,

Thanks for sharing! As you've noticed we have an on-going PR to set the maximum number of validation steps (and an issue: https://github.com/CAREamics/careamics/issues/480), so that's high priority for us.

The preferred way is to limit the number of steps per epoch, rather than perform intermediate validations, which indeed leads to more complications.

In general, we do try to leave the possibility for users to set all the Lightning parameters through our main API (`CAREamist` class). Note that you can also use the so-called ["Lightning API"](https://careamics.github.io/0.1/guides/lightning_api/), which is basically using the CAREamics classes in a Lightning context. This leaves in control of all the Lightning Trainer parameters! That's also an alternative for your use case.

Bottom line, the PR solution needs to be worked out but we should have a way to easily limit training steps next week. All Lightning Trainer parameters should still be accessible to users via the `CAREamist`, albeit a bit hidden. For full control, one can also use the Lightning API.

Now, I am interested in your use-case! You mentioned having 200 GB-big data, what format are you using? What type of data is it? 

I also want to point out that we are currently working on a next-generation file format compatible dataset. It is not fully features yet and not integrated (and the prediction writing directly to file on large dataset still need to be implemented), but it is [usable already](https://github.com/CAREamics/careamics/tree/main/src/careamics/dataset_ng/demos). Probably not directly useful, but maybe something to keep an eye on for you.

**Comment by gatoniel** [comment]
Hey, thanks for your response! I was not aware that I can also use the Lightning API. I will try that out!

Regarding the dataset: It is a nd2 file from a Nikon microscope. It is a multipoint file with 3d volumes. So the axis are `SZYX`. However, to be able to use your `predict_to_disk`, I exported the dataset into multiple `ZYX` tif volumes which are just lying in one folder without any further annotation.

**Comment by jdeschamps** [comment]
Should be implemented now thanks to https://github.com/CAREamics/careamics/pull/505.

It will takes a little bit of time to get a new version and the doc updated because we have some events to attend next week, but it'll come in September!
